src/network/networkd-sysctl.c

```diff
@@ -5,17 +5,189 @@
 #include "sysctl-util.h"
 
+#if HAVE_VMLINUX_H
+
+#include "bpf-link.h"
+
+#include "bpf/sysctl_monitor/sysctl-monitor-skel.h"
+#include "bpf/sysctl_monitor/sysctl-write-event.h"
+
+static struct sysctl_monitor_bpf *sysctl_monitor_bpf_free(struct sysctl_monitor_bpf *obj) {
+        sysctl_monitor_bpf__destroy(obj);
+        return NULL;
+}
+
+static struct ring_buffer *rb_free(struct ring_buffer *rb) {
+        sym_ring_buffer__free(rb);
+        return NULL;
+}
+
+DEFINE_TRIVIAL_CLEANUP_FUNC(struct sysctl_monitor_bpf *, sysctl_monitor_bpf_free);
+DEFINE_TRIVIAL_CLEANUP_FUNC(struct ring_buffer *, rb_free);
+
+static int sysct_write_event(void *ctx, void *data, size_t data_sz) {
+        struct sysctl_write_event *we = ASSERT_PTR(data);
+        Hashmap **sysctl_shadow = ASSERT_PTR(ctx);
+        _cleanup_free_ char *path = NULL;
+        char *value;
+
+        path = path_join("/proc/sys", we->name);
+        if (!path)
+                return log_oom();
+
+        /* If we never managed this handle, ignore it. */
+        value = hashmap_get(*sysctl_shadow, path);
+        if (!value)
+                return 0;
+
+        if (!streq(value, we->newvalue))
+                log_warning("'%s' changed sysctl '%s' from '%s' to '%s', we want '%s'", we->comm, we->name, we->current, we->newvalue, value);
+
+        return 0;
+}
+
+static int on_ringbuf_io(sd_event_source *s, int fd, uint32_t revents, void *userdata) {
+        struct ring_buffer *rb = ASSERT_PTR(userdata);
+        int r;
+
+        r = sym_ring_buffer__poll(rb, 1000);
+        if (r < 0 && errno != EINTR)
+                log_error_errno(errno, "Error polling ring buffer: %m");
+
+        return 0;
+}
+
+int sysctl_add_monitor(Manager *manager) {
+        _cleanup_(sysctl_monitor_bpf_freep) struct sysctl_monitor_bpf *obj = NULL;
+        _cleanup_(bpf_link_freep) struct bpf_link *sysctl_link = NULL;
+        _cleanup_(rb_freep) struct ring_buffer *sysctl_buffer = NULL;
+        _cleanup_close_ int cgroup_fd = -EBADF, rootcg = -EBADF;
+        int idx = 0;
+        int r;
+
+        assert(manager);
+
+        r = dlopen_bpf();
+        if (r < 0)
+                return log_info_errno(r, "sysctl monitor disabled, as BPF support is not available.");
+
+        rootcg = cg_path_open(SYSTEMD_CGROUP_CONTROLLER, "/");
+        if (rootcg < 0)
+                return log_warning_errno(rootcg, "Failed to open cgroup: %m");
+
+        obj = sysctl_monitor_bpf__open_and_load();
+        if (!obj)
+                return log_warning_errno(errno, "Failed to load sysctl monitor BPF program: %m");
+
+        cgroup_fd = cg_path_open(SYSTEMD_CGROUP_CONTROLLER, "/system.slice/systemd-networkd.service");
+        if (cgroup_fd < 0)
+                return log_warning_errno(cgroup_fd, "Failed to open cgroup: %m");
+
+        if (sym_bpf_map_update_elem(sym_bpf_map__fd(obj->maps.cgroup_map), &idx, &cgroup_fd, BPF_ANY))
+                return log_error_errno(errno, "Failed to update cgroup map: %m");
```

Comment:
logging is a bit all over the place. i.e. here you log at LOG_ERR level, above in LOG_WARN level, but the reaction to failure is otherwise the same, you fail this function.

usually we want that errors we ignore/handle gracefully are logged at LOG_WARN level, and those fatal for the operation we are executing are at LOG_ERR.